### PR TITLE
fix: restrict host and path of debugging URL

### PR DESCRIPTION
**Pull request by nic-chen**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

1. support configuring a list of APISIX gateway hosts in the config file
2. add an API to list the APISIX gateway hosts
3. filters the requests for debugging, and rejects directly if the host is not in the list. At the same time, prohibit  sensitive paths that start with `/apisix/`

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

**Comment by liuxiran** [comment]
FE e2e problem can be fixed by #1895

**Comment by nic-chen** [comment]
@liuxiran  seems we need to update FE test cases due to the host and path restriction.

File: api/conf/conf.yaml [hunk 1]
```diff
@@ -59,6 +59,9 @@ conf:
                          # log example: 2020-12-09T16:38:09.039+0800	INFO	filter/logging.go:46	/apisix/admin/routes/r1	{"status": 401, "host": "127.0.0.1:9000", "query": "asdfsafd=adf&a=a", "requestId": "3d50ecb8-758c-46d1-af5b-cd9d1c820156", "latency": 0, "remoteIP": "127.0.0.1", "method": "PUT", "errs": []}
   max_cpu: 0             # supports tweaking with the number of OS threads are going to be used for parallelism. Default value: 0 [will use max number of available cpu cores considering hyperthreading (if any)]. If the value is negative, is will not touch the existing parallelism profile.
 
+  gateways:              # host list of APISIX gateway, currently only used as a list of debuggable hosts
```
Comment (tokers) [inline api/conf/conf.yaml:62]: Better to add big hints for users to be careful about this field.

File: api/internal/handler/route_online_debug/route_online_debug.go [hunk 2]
```diff
@@ -92,12 +108,60 @@ func (h *Handler) DebugRequestForwarding(c droplet.Context) (interface{}, error)
...
+	if len(conf.Gateways) < 1 {
+		return errors.New("host list of APISIX gateways not configured")
+	}
+	for _, gatewayHost := range conf.Gateways {
+		if host == gatewayHost {
+			return nil
+		}
+	}
+
+	return errors.New("doesn't match any host of APISIX gateways")
```
Comment (tokers) [inline api/internal/handler/route_online_debug/route_online_debug.go:121]:
```suggestion
	return errors.New("host denied")
```

File: api/internal/handler/route_online_debug/route_online_debug.go [hunk 3]
```diff
@@ -92,12 +108,60 @@ func (h *Handler) DebugRequestForwarding(c droplet.Context) (interface{}, error)
 type HTTPProtocolSupport struct {
 }
 
+func checkHost(host string) error {
```
Comment (tokers) [inline api/internal/handler/route_online_debug/route_online_debug.go:111]: Return a boolean value is better.

File: api/conf/conf.yaml (hunk 1 above)
Comment (gxthrj) [reply api/conf/conf.yaml:62]:
```suggestion
  debug_gateways:              # host list of APISIX gateway, currently only used as a list of debuggable hosts
```

File: api/internal/handler/route_online_debug/route_online_debug.go [hunk 4]
```diff
@@ -92,12 +108,60 @@ func (h *Handler) DebugRequestForwarding(c droplet.Context) (interface{}, error)
...
+			return nil
+		}
+	}
+
+	return errors.New("doesn't match any host of APISIX gateways")
+}
+
+func checkPath(path string) error {
+	if strings.HasPrefix(path, "/apisix/") {
+		return errors.New("the path is forbidden for debugging")
```
Comment (gxthrj) [inline api/internal/handler/route_online_debug/route_online_debug.go:126]:
```suggestion
		return fmt.Errorf(("The path %s is forbidden for debugging", path)
```

File: api/internal/handler/route_online_debug/route_online_debug.go [hunk 5]
```diff
@@ -92,12 +108,60 @@ func (h *Handler) DebugRequestForwarding(c droplet.Context) (interface{}, error)
...
+	}
+	return nil
+}
+
+func checkMethod(method string) error {
+	if exists := allowMethods[strings.ToUpper(method)]; !exists {
+		return errors.New("the method is not allowed for debugging")
```
Comment (gxthrj) [inline api/internal/handler/route_online_debug/route_online_debug.go:133]:
```suggestion
		return fmt.Errorf("The method %s is not allowed for debugging", method)
```

File: api/internal/handler/route_online_debug/route_online_debug.go [hunk 6]
```diff
@@ -92,12 +108,60 @@ func (h *Handler) DebugRequestForwarding(c droplet.Context) (interface{}, error)
...
 
+func checkHost(host string) error {
+	if len(conf.Gateways) < 1 {
+		return errors.New("host list of APISIX gateways not configured")
+	}
+	for _, gatewayHost := range conf.Gateways {
+		if host == gatewayHost {
+			return nil
+		}
+	}
```
Comment (starsz) [inline api/internal/handler/route_online_debug/route_online_debug.go:119]: Move the `inArray` in `ssl.go` to `utils.go` and use it.

File: api/internal/handler/tool/tool.go [hunk 7]
```diff
@@ -109,3 +111,7 @@ func (h *Handler) VersionMatch(c droplet.Context) (interface{}, error) {
 
 	return &output, nil
 }
+
+func (h *Handler) Gateways(_ droplet.Context) (interface{}, error) {
+	return conf.Gateways, nil
```
Comment (starsz) [inline api/internal/handler/tool/tool.go:116]: Miss e2e test to test it?

**Pull request closed without merging**